test/mpi/test_mpi_cuda.c

```diff
@@ -0,0 +1,695 @@
+
+    pop_and_release_primary_context(params->cu_dev);
+
+    return res;
+}
+
+static int test_alloc_prim_send_no(const test_params_t *params)
+{
+    alloc_mem_t alloc_mem_send, alloc_mem_recv;
+    CUcontext primary_ctx;
+    int res;
+
+    retain_and_push_primary_context(params->cu_dev);
+
+    alloc_mem(params, &alloc_mem_send, &alloc_mem_recv);
+
+    CUDA_CHECK(cuCtxPopCurrent(&primary_ctx));
+    check_no_current_context();
+
+    mpi_pingpong(params->rank, alloc_mem_send.ptr, alloc_mem_recv.ptr,
+                 params->size);
+
+    CUDA_CHECK(cuCtxPushCurrent(primary_ctx));
+
+    res = check_and_free_mem(params, &alloc_mem_send, &alloc_mem_recv);
+
+    pop_and_release_primary_context(params->cu_dev);
+
+    return res;
+}
+
+static int test_alloc_prim_send_thread(const test_params_t *params)
+{
+    alloc_mem_t alloc_mem_send, alloc_mem_recv;
+    pthread_t thread;
+    thread_arg_t thread_arg;
+    int res;
+
+    retain_and_push_primary_context(params->cu_dev);
+
+    alloc_mem(params, &alloc_mem_send, &alloc_mem_recv);
+
+    thread_arg.rank     = params->rank;
+    thread_arg.send_ptr = alloc_mem_send.ptr;
+    thread_arg.recv_ptr = alloc_mem_recv.ptr;
+    thread_arg.size     = params->size;
+    pthread_create(&thread, NULL, mpi_pingpong_thread, &thread_arg);
+    pthread_join(thread, NULL);
+
+    res = check_and_free_mem(params, &alloc_mem_send, &alloc_mem_recv);
+
+    pop_and_release_primary_context(params->cu_dev);
+
+    return res;
+}
+
+static int test_alloc_prim_send_user(const test_params_t *params)
+{
+    alloc_mem_t alloc_mem_send, alloc_mem_recv;
+    CUcontext primary_ctx, user_ctx;
+    int res;
+
+    retain_and_push_primary_context(params->cu_dev);
+
+    alloc_mem(params, &alloc_mem_send, &alloc_mem_recv);
+
+    CUDA_CHECK(cuCtxPopCurrent(&primary_ctx));
+    CUDA_CHECK(cuCtxCreate(&user_ctx, 0, params->cu_dev));
+
+    mpi_pingpong(params->rank, alloc_mem_send.ptr, alloc_mem_recv.ptr,
+                 params->size);
+
+    CUDA_CHECK(cuCtxDestroy(user_ctx));
+    CUDA_CHECK(cuCtxPushCurrent(primary_ctx));
+
+    res = check_and_free_mem(params, &alloc_mem_send, &alloc_mem_recv);
+
+    pop_and_release_primary_context(params->cu_dev);
+
+    return res;
+}
+
+static int test_alloc_user_send_prim(const test_params_t *params)
+{
+    CUcontext user_ctx;
+    alloc_mem_t alloc_mem_send, alloc_mem_recv;
+    int res;
+
+    CUDA_CHECK(cuCtxCreate(&user_ctx, 0, params->cu_dev));
+
+    alloc_mem(params, &alloc_mem_send, &alloc_mem_recv);
+
+    retain_and_push_primary_context(params->cu_dev);
+    mpi_pingpong(params->rank, alloc_mem_send.ptr, alloc_mem_recv.ptr,
+                 params->size);
+    pop_and_release_primary_context(params->cu_dev);
+
+    res = check_and_free_mem(params, &alloc_mem_send, &alloc_mem_recv);
+
+    CUDA_CHECK(cuCtxDestroy(user_ctx));
+
+    return res;
+}
+
+static int test_alloc_prim_send_other_prim(const test_params_t *params)
```

Comment:
can we do send/recv between all possible combinations of available gpus?

Reply:
Yes, we can. However, it doesn't bring any additional value for the test. The point is to test the following scenario:
```
1. set device X
2. allocate data
3. set device Y
4. send data
5. unset device Y
6. free data
7. unset device X
```

Reply:
considering that we may want to run this test also on non-CI machines (possibly DGX-like), running all-2-all  seems to be a better coverage. wdyt?

Reply:
I don't think that this will improve coverage, or I don't see the case. Could you please explain the scenario that will be covered by these changes.

Reply:
some configurations, like DGX v1 may have different communication properties between different GPUs
you can add todo for later addition for now